File: language-testing/tcp_echo_server/tcp_echo_server.rb
#! /usr/bin/env ruby

require 'socket'

def start_echo_server port
  socky = TCPServer.new port
  loop do
    client = socky.accept
    data = client.gets
    client.puts data
    client.close

    File.open "tcp_echo_server.log", "a" do |f|
      f.puts data
    end
  end
end

def die error
  $stderr.puts "Usage: #{$0} port\n#{error}"
  exit 1
end

if __FILE__ == $0
  die "No port given!" unless ARGV.length > 0
  die "Error: Port contains non-digits!" unless ARGV[0] =~ /^\d+$/
  start_echo_server ARGV[0].to_i
end

File: language-testing/tcp_echo_server/tcp_echo_server.pl
#! /usr/bin/env perl

use strict;
use warnings;

use IO::Socket;
$| = 1;

sub start_echo_server {
  my $sock = new IO::Socket::INET (
    LocalPort => $_[0],
    Proto => "tcp",
    Listen => 5,
    Reuse => 1
  ) or die "Failed to start socket: $!\n";

  for (;;) {
    my $client = $sock->accept() or die "Accept: $!\n";
    my $data = <$client>;
    print $client $data;
    close $client;
    open my $f, ">>", "tcp_echo_server.log" or die "open: $!\n";
    print $f $data;
    close $f;
  }
}

if ($#ARGV == 0) {
  &start_echo_server($ARGV[0]);
} else {
  print "Usage: $0 port\n";
  exit 1;
}
